src/software-device.cpp

```diff
@@ -0,0 +1,126 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2015 Intel Corporation. All Rights Reserved.
+
+#include "software-device.h"
+#include "stream.h"
+
+namespace librealsense
+{
+    software_device::software_device()
+        : device(std::make_shared<context>(backend_type::standard), {})
+    {
+        register_info(RS2_CAMERA_INFO_NAME, "Bypass-Sensor");
+    }
+
+    software_sensor& software_device::add_software_sensor(const std::string& name)
+    {
+        auto sensor = std::make_shared<software_sensor>(name, this);
+        add_sensor(sensor);
+        _software_sensors.push_back(sensor);
+
+        return *sensor;
+    }
+
+    software_sensor& software_device::get_software_sensor(int index)
+    {
+        return *_software_sensors[index];
+    }
+
+    void software_device::set_matcher_type(rs2_matchers matcher)
+    {
+        _matcher = matcher;
+    }
+
+    software_sensor::software_sensor(std::string name, software_device* owner)
+        : sensor_base(name, owner)
+    {
+
+    }
+
+    std::shared_ptr<matcher> software_device::create_matcher(const frame_holder& frame) const
+    {
+        std::vector<stream_interface*> profiles;
+
+        for (auto&& s : _software_sensors)
+            for (auto&& p : s->get_stream_profiles())
+                profiles.push_back(p.get());
+
+        return matcher_factory::create(_matcher, profiles);
+    }
+
+    void software_sensor::add_video_stream(rs2_video_stream video_stream)
+    {
+        auto profile = std::make_shared<video_stream_profile>(
+            platform::stream_profile{ (uint32_t)video_stream.width, (uint32_t)video_stream.height, (uint32_t)video_stream.fps, 0 });
+        profile->set_dims(video_stream.width, video_stream.height);
+        profile->set_format(video_stream.fmt);
+        profile->set_framerate(video_stream.fps);
+        profile->set_stream_index(video_stream.index);
+        profile->set_stream_type(video_stream.type);
+        profile->set_unique_id(video_stream.uid);
+        profile->set_intrinsics([=]() {return video_stream.intrinsics; });
+        _profiles.push_back(profile);
```

Comment:
What will happen if the user adds the same stream (e.g Depth) twice ?
Is the specific stream type will be replaced/cloned/raise an exception?